src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs

```diff
@@ -154,10 +154,31 @@ public IList<FailurePoint> FailurePoints
         /// <summary>
         /// Compares two objects for equality within a tolerance.
         /// </summary>
-        public bool AreEqual(object x, object y, ref Tolerance tolerance, bool topLevelComparison = true)
+        public bool AreEqual(object x, object y, ref Tolerance tolerance)
+        {
+            return AreEqual(x, y, ref tolerance, true);
+        }
+
+        /// <summary>
+        /// Compares two objects for equality within a tolerance.
+        /// </summary>
+        public bool AreEqual(object x, object y, ref Tolerance tolerance, bool topLevelComparison)
+        {
+            return AreEqual(x, y, ref tolerance, new ComparisonState()
+            {
+                TopLevelComparison = topLevelComparison,
+                Comparisons = new List<ComparisonState.Comparison>()
+            });
+        }
+
+        internal bool AreEqual(object x, object y, ref Tolerance tolerance, ComparisonState state)
```

Comment:
Note to reviewer: All the internal comparer signatures were updated to receive a `ComparisonState` object instead of a `bool`. `ComparisonState` is an internal struct which tracks, among other things, the existing "topLevelComparison" state as well as a list of objects already compared to each other.

To ensure backwards-compatibiliity, the public API of NUnitEqualityComparer was left as-is, with an internal overload added to receive this new object.